proposals/0022-multi-stake.md

```diff
@@ -0,0 +1,178 @@
+
+Small stakes need to be delegated, while not causing problems in the validator's
+stakes cache.
+
+Side note: this isn't my first proposal for a new stake account type. Here's an earlier
+attempt meant for redelegation: https://github.com/solana-labs/solana/pull/24762
+
+## Alternatives Considered
+
+The easiest solution is to change the validator's stakes cache, since that only
+impacts the specific implementation in the Solana Labs validator.
+
+In the perfect situation, small stakes are properly delegated, but don't receive
+rewards. Meaning:
+
+* they don't take up space in the stakes cache
+* they take the correct amount of epoch boundaries to activate / deactivate
+* they are included in a validator's voting power
+
+There are a few ways of solving this at the stakes cache level, and they all have
+serious issues.
+
+* Include small stakes in the cache, but don't pay out rewards
+
+This is the simplest solution, which essentially changes "minimum delegation amount"
+to mean "minimum reward earning amount". So you can delegate stake accounts with
+less than the minimum delegation, but they won't earn rewards.
+
+Unfortunately, small stakes can still bloat the stakes cache from a memory level.
+
+* Don't include small stakes in the cache, but track them in transactions
+
+In this solution, on every transaction, if the runtime sees a small stake, it
+includes its delegation amount in the validator's voting power, but doesn't
+add its pubkey and `Delegation` to the cache.
+
+The runtime tracks the pre and post state of every account, and for a successful
+transaction, subtracts the old small stake delegation amount, and adds the new small
+stake delegation amount.
+
+During rewards payout, small stakes are completely omitted, since they are not
+present in the cache.
+
+This is incredibly brittle, however, and introduces more overhead in the runtime.
+The bank must hold onto pre-states for all transactions to debit the stakes cache. And,
+if called from the outside, all `store_accounts` functions on the bank could easily
+invalidate the stakes cache.
+
+For example, if you store a small stake directly, what does the cache do? What
+was the pre-state of that small stake? There's no way to know.
+
+## Detailed Design
+
+Rather than hacking the validator, let's change the stake program.
+
+With a new "multi stake" account, carrying *two* more `Delegation`s, the stake
+program becomes much more flexible.
+
+During initial delegation, the "multi stake" account must have enough to cover
+the minimum stake delegation amount. The two other `Delegation` instances may be
+used for small stake movements.
+
+Here are some example uses:
+
+### Add or remove small stake
+
+To add a small stake to your account, transfer the lamports, call `delegate` once
+more, and now your stake account has two delegations: the initial one, and the
+newly activating one.
+
+Once the new `Delegation` is active, a permissionless `update` instruction follows
```

Comment:
The idea of "update" is for intra-account merges and clearing old delegation instances. For example, if the account is delegated to validator A, and I want to add a little more stake, I transfer lamports and activate a second delegation. Once that's active, let's say I want to add lamports again. Both delegations in the account are currently active, so I should be able to merge those two.

The current `merge` instruction isn't particularly suited to doing that since it assumes draining the source account, but we could update "merge" to work with the same account passed in twice.

And sorry, I don't get the "we should just be able to wrap the Delegations in a container with the relevant Delegation methods implemented" part. Can you give a quick example of what one of those functions would look like?

Reply:
> The idea of "update" is for intra-account merges and clearing old delegation instances. For example, if the account is delegated to validator A, and I want to add a little more stake, I transfer lamports and activate a second delegation. Once that's active, let's say I want to add lamports again. Both delegations in the account are currently active, so I should be able to merge those two.
> 
> The current `merge` instruction isn't particularly suited to doing that since it assumes draining the source account, but we could update "merge" to work with the same account passed in twice.
> 
> Unless you're thinking that we keep all of the same instructions and have them work with the same account passed in twice.

nah i was thinking it's pretty cheap to merge delegations and we could just try any time we need to operate on the secondary (or tertiary) delegations.

> And sorry, I don't get the "we should just be able to wrap the Delegations in a container with the relevant Delegation methods implemented" part. Can you give a quick example of what one of those functions would look like?

was thinking stuff along the lines of...
```
struct Delegations([Delegation; 2]);

impl Delegations {
    pub fn get_active_mut(&mut self, epoch: Epoch) -> Option<&mut Delegation> {
        self.0
            .iter_mut()
            .find(|delegation| {
                epoch > delegation.activation_epoch &&
                    epoch < delegation.deactivation_epoch
            })
    }

    pub fn get_transient_mut(&mut self, epoch: Epoch) -> Option<&mut Delegation> {
        self.0
            .iter_mut()
            .find(|delegation| {
                // never delegated
                (delegation.activation_epoch == 0 && delegation.deactivation_epoch == Epoch::MAX) ||
                // activating
                (epoch == deleation.activation_epoch && delegation.deactivation_epoch == Epoch::MAX) ||
                // deactivating
                (epoch == delegation.deactivation_epoch) ||
                // deactivated
                (epoch > delegation.deactivation_epoch)
            })
    }

    pub fn stake_activating_deactivating(
        &self,
        target_epoch: Epoch,
        history: Option<&StakeHistory>,
    ) -> StakeActivationStatus {
        self.0.iter()
            .fold(
                StakeActivationStatus::default(),
                |acc, delegation| {
                    acc += delegation
                        .stake_activating_deactivating(
                            target_epoch,
                            history,
                        )
                })
    }
}
```

though realizing here that we really need to duplicate the `Stake` struct.  i forgot that that's where we tracked observed credits for w/e reason

Reply:
> nah i was thinking it's pretty cheap to merge delegations and we could just try any time we need to operate on the secondary (or tertiary) delegations.

I worry about being too clever on stuff like that and breaking things, but I'll certainly look into it.

> was thinking stuff along the lines of...

Ok yeah, perfect, I'm pretty sure this is vaguely what @mvines suggested too